### PR TITLE
VxDesign: Ballot layout density option

### PR DESCRIPTION
**Pull request by jonahkagan**
## Overview

Adds a ballot layout density option to VxDesign. There are three density levels to choose from. Increasing the density changes the font size and spacing to help fit more ballot content on each page, while still maintaining bubbles located on grid intersections.

## Demo Video or Screenshot

https://github.com/votingworks/vxsuite/assets/530106/d3f4820d-cefc-47b2-9c1b-93a6300a0703

## Testing Plan
- Visual inspection of ballots
- Updated automated testing of interpretation

## Checklist

- [ ] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.

- [ ] I have added a screenshot and/or video to this PR to demo the change
- [ ] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates

### CODE REVIEW

File: apps/design/frontend/src/contests_screen.tsx [hunk 1]
```diff
@@ -391,6 +391,7 @@ function ContestForm({
                         <Input
                           type="text"
                           value={candidate.name}
+                          autoFocus
```
Comment (jonahkagan) [inline apps/design/frontend/src/contests_screen.tsx:394]: Unrelated change to make it a bit quicker to type in candidate names in sequence

File: apps/design/shared/jest.config.js [hunk 2]
```diff
@@ -14,5 +14,6 @@ module.exports = {
       lines: 0,
     },
   },
+  coverageProvider: 'v8',
```
Comment (jonahkagan) [inline apps/design/shared/jest.config.js:17]: Unrelated change due to getting the "heap out of memory" error

File: apps/design/shared/src/layout.ts [hunk 3]
```diff
@@ -104,6 +80,58 @@ function textWrap(text: string, fontStyle: FontStyle, width: number): string[] {
...
 
+function textWrap(text: string, fontStyle: FontStyle, width: number): string[] {
+  return text.split('\n').flatMap((line) => wrapLine(line, fontStyle, width));
+}
+
+function textHeight(textLines: string[], fontStyle: FontStyle): number {
+  return textLines.length * fontStyle.lineHeight;
+}
+
+function TextBlock({
```
Comment (jonahkagan) [inline apps/design/shared/src/layout.ts:91]: Added this new subcomponent to help with rendering multi-line text where the lines have different styling

File: apps/design/shared/src/layout.ts [hunk 4]
```diff
@@ -682,10 +747,50 @@ function CandidateContest({
   const optionTextAlign =
     bubblePosition === BallotTargetMarkPosition.Left ? 'left' : 'right';
 
-  const optionRowHeight = 2;
```
Comment (jonahkagan) [inline apps/design/shared/src/layout.ts:685]: The next bit of diff adds text wrapping for candidate names, dynamically computing the height needed for each contest option.

File: apps/design/shared/src/layout.ts
Comment (jonahkagan) [inline apps/design/shared/src/layout.ts:1]: Note that I'm still planning a larger rewrite of layout, and this feature will probably change, so I wouldn't spend too much time in the nitty gritty of it.

File: apps/design/shared/src/layout.ts (hunk 3 above)
Comment (adghayes) [reply apps/design/shared/src/layout.ts:91]: Cool, looks like this helps simplify a lot below!

File: apps/design/shared/src/layout.ts [hunk 5]
```diff
@@ -1348,7 +1428,9 @@ function layOutBallotHelper(
           m,
         });
         if (height > gridHeight(m.MAX_CONTEST_ROW_HEIGHT, m)) {
-          throw new Error(`Contest ${contest.id} is too tall to fit on a page`);
+          throw new Error(
+            `Contest is too tall to fit on a page: ${contest.title} `
+          );
```
Comment (adghayes) [inline apps/design/shared/src/layout.ts:1433]: longer-term I wonder if we could automatically attempt different densities to try to fit the large contests

**Review by adghayes — APPROVED** [review]
LGTM

File: apps/design/shared/src/layout.ts (hunk 5 above)
Comment (jonahkagan) [reply apps/design/shared/src/layout.ts:1433]: Something in that direction could be cool!
